src/modules/Dashboard/components/ActivityGrid/ActivityGrid.tsx

```diff
@@ -0,0 +1,126 @@
+  isLoading: boolean;
+  'data-testid': string;
+};
+
+export const ActivityGrid = ({
+  rows,
+  isLoading = false,
+  'data-testid': dataTestid,
+}: ActivitiesGridProps) => {
+  const navigate = useNavigate();
+  const { t } = useTranslation('app');
+  const { appletId, participantId: _participantId } = useParams();
+
+  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
+
+  const isEmpty = !rows?.length && !isLoading;
+
+  const getEmptyComponent = () => {
+    if (isEmpty) {
+      return appletId ? t('noActivitiesForApplet') : t('noActivities');
+    }
+  };
+
+  return (
+    <StyledBody>
+      {isLoading && <Spinner />}
+
+      <StyledFlexWrap sx={{ gap: 1.2, mb: 2.4 }}>
+        {appletId && (
+          <>
+            <StyledFlexTopCenter sx={{ gap: 1.2 }}>
+              <StyledButton
+                variant="outlined"
+                startIcon={<Svg width={18} height={18} id="slider-rows" />}
+                // TODO: Implement filters
+                // https://mindlogger.atlassian.net/browse/M2-5530
+                onClick={() => alert('TODO: filters')}
+                data-testid={`${dataTestid}-filters`}
+              >
+                {t('filters')}
+              </StyledButton>
+
+              <ButtonWithMenu
+                variant="outlined"
+                label={t('sortBy')}
+                anchorEl={anchorEl}
+                setAnchorEl={setAnchorEl}
+                // TODO: Implement sorting
+                // https://mindlogger.atlassian.net/browse/M2-5445
+                menuItems={[
+                  {
+                    title: 'TODO: Sort options',
+                    action: () => alert('TODO: Sort options'),
+                  },
+                ]}
+                startIcon={<></>}
+                data-testid={`${dataTestid}-sort-by`}
+              />
+            </StyledFlexTopCenter>
+
+            <StyledFlexWrap sx={{ gap: 1.2, ml: 'auto' }}>
+              <StyledButton
+                // TODO: Replace with missing `tonal` button variant as shown in Figma
+                // https://mindlogger.atlassian.net/browse/M2-6071
+                variant="outlined"
+                // TODO: Implement assign
+                // https://mindlogger.atlassian.net/browse/M2-5710
+                onClick={() => alert('TODO: Assign activity')}
+                data-testid={`${dataTestid}-assign`}
+                sx={{ minWidth: '10rem' }}
+              >
+                {t('assign')}
+              </StyledButton>
+
+              <StyledButton
+                variant="contained"
+                onClick={() => navigate(generatePath(page.builderAppletActivities, { appletId }))}
+                data-testid={`${dataTestid}-add-activity`}
+                sx={{ minWidth: '13.2rem' }}
+              >
+                {t('addActivity')}
+              </StyledButton>
+            </StyledFlexWrap>
+          </>
+        )}
+      </StyledFlexWrap>
+
+      {rows?.length && (
```

Comment:
Protect against `rows.length === 0`:
```suggestion
      {!!rows?.length && (
```